File: backend/src/main/scala/kpn/api/common/PoiAnalysis.scala
package kpn.api.common

import kpn.api.custom.Tag

case class PoiAnalysis(
  layers: Seq[String] = Seq.empty,
  mainTags: Seq[Tag] = Seq.empty,
  extraTags: Seq[Tag] = Seq.empty,
  name: Option[String] = None,
  subject: Option[String] = None,
  description: Option[String] = None,
  addressLine1: Option[String] = None,
  addressLine2: Option[String] = None,
  phone: Option[String] = None,
  email: Option[String] = None,
  fax: Option[String] = None,
  facebook: Option[String] = None,
  twitter: Option[String] = None,
  website: Option[String] = None,
  wikidata: Option[String] = None,
  wikipedia: Option[String] = None,
  molenDatabase: Option[String] = None,
  hollandscheMolenDatabase: Option[String] = None,
  image: Option[String] = None,
  imageLink: Option[String] = None,
  imageThumbnail: Option[String] = None,
  mapillary: Option[String] = None,
  wheelchair: Option[String] = None,
  onroerendErfgoed: Option[String] = None,
  openingHours: Option[String] = None,
  serviceTimes: Option[String] = None,
  cuisine: Option[String] = None,
  denomination: Option[String] = None
) {
  def hasImage: Boolean = {
    image.isDefined ||
      imageLink.isDefined ||
      imageThumbnail.isDefined ||
      mapillary.isDefined
  }

  def hasLink: Boolean = {
    facebook.isDefined ||
      twitter.isDefined ||
      website.isDefined ||
      wikidata.isDefined ||
      wikipedia.isDefined ||
      molenDatabase.isDefined ||
      hollandscheMolenDatabase.isDefined ||
      onroerendErfgoed.isDefined
  }

  def address: Option[String] = {
    addressLine1 match {
      case None => addressLine2
      case Some(addressLine1Value) =>
        addressLine2 match {
          case Some(addressLine2Value) => Some(s"$addressLine1Value, $addressLine2Value")
          case None => addressLine1
        }
    }
  }

  def nameDescription: Option[String] = {
    name match {
      case Some(nameValue) => Some(nameValue)
      case None => description
    }
  }
}

File: backend/api/src/main/java/kpn/api/common/PoiAnalysis.java
package kpn.api.common;

import kpn.api.custom.Tag;

import java.util.Optional;
import com.google.common.collect.ImmutableList;

public record PoiAnalysis(
  ImmutableList<String> layers,
  ImmutableList<Tag> mainTags,
  ImmutableList<Tag> extraTags,
  Optional<String> name,
  Optional<String> subject,
  Optional<String> description,
  Optional<String> addressLine1,
  Optional<String> addressLine2,
  Optional<String> phone,
  Optional<String> email,
  Optional<String> fax,
  Optional<String> facebook,
  Optional<String> twitter,
  Optional<String> website,
  Optional<String> wikidata,
  Optional<String> wikipedia,
  Optional<String> molenDatabase,
  Optional<String> hollandscheMolenDatabase,
  Optional<String> image,
  Optional<String> imageLink,
  Optional<String> imageThumbnail,
  Optional<String> mapillary,
  Optional<String> wheelchair,
  Optional<String> onroerendErfgoed,
  Optional<String> openingHours,
  Optional<String> serviceTimes,
  Optional<String> cuisine,
  Optional<String> denomination
) {}

/* TODO migrate

  def hasImage: Boolean = {
    image.isDefined ||
      imageLink.isDefined ||
      imageThumbnail.isDefined ||
      mapillary.isDefined
  }

  def hasLink: Boolean = {
    facebook.isDefined ||
      twitter.isDefined ||
      website.isDefined ||
      wikidata.isDefined ||
      wikipedia.isDefined ||
      molenDatabase.isDefined ||
      hollandscheMolenDatabase.isDefined ||
      onroerendErfgoed.isDefined
  }

  def address: Option[String] = {
    addressLine1 match {
      case None => addressLine2
      case Some(addressLine1Value) =>
        addressLine2 match {
          case Some(addressLine2Value) => Some(s"$addressLine1Value, $addressLine2Value")
          case None => addressLine1
        }
    }
  }

  def nameDescription: Option[String] = {
    name match {
      case Some(nameValue) => Some(nameValue)
      case None => description
    }
  }

*/
